lib/src/gitattributes.rs

```diff
@@ -0,0 +1,705 @@
+                (val.name.as_str().to_string(), val.state)
+            })
+            .collect::<HashMap<String, State>>();
+
+        // go over attributes and mark as unspecified if not set
+        for attribute_name in attribute_names.as_ref() {
+            map.entry(attribute_name.to_string())
+                .or_insert(State::Unspecified);
+        }
+
+        Ok(map)
+    }
+
+    fn get_git_attributes_node(&self, path: &RepoPath) -> Arc<GitAttributesNode> {
+        let mut val = self.node_cache.lock().expect("Not be poisoned");
+        self.inner(&mut val, path)
+    }
+
+    fn inner(
+        &self,
+        map: &mut HashMap<RepoPathBuf, Arc<GitAttributesNode>>,
+        path: &RepoPath,
+    ) -> Arc<GitAttributesNode> {
+        if let Some(node) = map.get(path) {
+            return node.clone();
+        }
+        let parent = path.parent().map(|parent| self.inner(map, parent));
+        let new_node = Arc::new(GitAttributesNode {
+            disk_first: OnceCell::new(),
+            store_first: OnceCell::new(),
+            disk_file_loader: self.disk_file_loader.clone(),
+            store_file_loader: self.store_file_loader.clone(),
+            parent,
+            path: path.to_owned(),
+        });
+        map.insert(path.to_owned(), new_node.clone());
+        new_node
+    }
+}
+
+/// Errors for GitAttributes
+#[derive(Debug, thiserror::Error)]
+#[error("{message}")]
+pub struct GitAttributesError {
+    message: String,
+    #[source]
+    source: Box<dyn std::error::Error + Send + Sync>,
+}
+
+type TestStore = HashMap<RepoPathBuf, Result<String, String>>;
+
+#[async_trait::async_trait]
+impl FileLoader for TestStore {
+    async fn load(
+        &self,
+        path: &RepoPath,
+    ) -> Result<Option<Box<dyn AsyncRead + Send + Unpin>>, GitAttributesError> {
+        let Some(mocked_result) = self.get(path) else {
+            return Ok(None);
+        };
+        let data = mocked_result.clone(); // can't clone?
+        match data {
+            Ok(data) => Ok(Some(Box::new(Cursor::new(Vec::from(data))))),
+            Err(message) => Err(GitAttributesError {
+                message: message.clone(),
+                source: message.into(),
+            }),
+        }
+    }
+}
+#[cfg(test)]
+mod tests {
+
+    use gix_attributes::state::Value;
+
+    use super::*;
+
+    fn repo_path(path: &str) -> &RepoPath {
+        RepoPath::from_internal_string(path).unwrap()
+    }
+
+    fn create_git_attributes(files: &[(&'static str, &'static str)]) -> GitAttributes {
+        let data: TestStore = files
+            .iter()
+            .map(|(key, value)| {
+                (
+                    RepoPathBuf::from_internal_string(key.to_string()).unwrap(),
+                    Ok(value.to_string()),
+                )
+            })
+            .collect();
+
+        GitAttributes::new(HashMap::new(), data)
+    }
+
+    // macros so tests fail on the right line
```

Comment:
Can we just use a function instead of macros?

If we use functions, and the test fails, the entire stack will be printed, so we don't lose the precise line where the assertion fails. If we use macros, we only see the line where the macro is used, we won't able to identify the exact line in macro where the panic happens.

In addition, functions have better readability than macro.